File: Modules/ServerTypes/Administration/UserActivityHistoryForm.ts
import { StringEditor, DateEditor, TextAreaEditor, PrefixedContext, initFormType } from "@serenity-is/corelib";

export interface UserActivityHistoryForm {
    UserId: StringEditor;
    Username: StringEditor;
    ActivityType: StringEditor;
    ActivityTime: DateEditor;
    IpAddress: StringEditor;
    Location: StringEditor;
    Isp: StringEditor;
    Timezone: StringEditor;
    UserAgent: TextAreaEditor;
    SessionId: StringEditor;
    Details: TextAreaEditor;
}

export class UserActivityHistoryForm extends PrefixedContext {
    static readonly formKey = 'Administration.UserActivityHistory';
    private static init: boolean;

    constructor(prefix: string) {
        super(prefix);

        if (!UserActivityHistoryForm.init)  {
            UserActivityHistoryForm.init = true;

            var w0 = StringEditor;
            var w1 = DateEditor;
            var w2 = TextAreaEditor;

            initFormType(UserActivityHistoryForm, [
                'UserId', w0,
                'Username', w0,
                'ActivityType', w0,
                'ActivityTime', w1,
                'IpAddress', w0,
                'Location', w0,
                'Isp', w0,
                'Timezone', w0,
                'UserAgent', w2,
                'SessionId', w0,
                'Details', w2
            ]);
        }
    }
}

File: Modules/Administration/UserActivityHistory/UserActivityHistoryForm.cs
using Serenity.ComponentModel;

namespace SerenityProjem.Administration;

[FormScript("Administration.UserActivityHistory")]
[BasedOnRow(typeof(UserActivityHistoryRow), CheckNames = true)]
public class UserActivityHistoryForm
{
    [Tab("General")]
    public string UserId { get; set; }
    public string Username { get; set; }
    public string ActivityType { get; set; }
    public DateTime ActivityTime { get; set; }
    
    [Tab("Location & Device")]
    public string IpAddress { get; set; }
    public string Location { get; set; }
    public string Isp { get; set; }
    public string Timezone { get; set; }
    
    [Tab("Technical")]
    [TextAreaEditor(Rows = 3)]
    public string UserAgent { get; set; }
    public string SessionId { get; set; }
    
    [TextAreaEditor(Rows = 5)]
    public string Details { get; set; }
}